### PR TITLE
(PUP-10091) Add FacterNg dependency

### PR DESCRIPTION
**Pull request by sebastian-miclea**

### CODE REVIEW

**Comment by sebastian-miclea** [comment]
Closed and reopened with independent ticket: https://github.com/puppetlabs/puppet/pull/7765

File: Gemfile [hunk 1]
```diff
@@ -14,6 +14,7 @@ end
 
 # override .gemspec deps - may issue warning depending on Bundler version
 gem "facter", *location_for(ENV['FACTER_LOCATION']) if ENV.has_key?('FACTER_LOCATION')
+gem "facter-ng"
```
Comment (joshcooper) [inline Gemfile]: Could you add the `@location_for` logic that way we can point it to a local checkout of facter-ng?

File: lib/puppet/type/group.rb [hunk 2]
```diff
@@ -125,7 +125,7 @@ def actual_should(currentvalue, newvalue)
         newvalue = munge_members_value(newvalue)
 
         if @resource[:auth_membership]
-          newvalue.uniq.sort 
+          newvalue.uniq.sort
```
Comment (joshcooper) [inline lib/puppet/type/group.rb]: Can you revert the unrelated whitespace changes?

**Comment by ciprianbadescu** [comment]
lgtm

File: lib/puppet.rb [hunk 3]
```diff
@@ -9,7 +9,7 @@
 # see the bottom of the file for further inclusions
 # Also see the new Vendor support - towards the end
 #
-require 'facter'
+require ENV['USE_FACTER_NG'] ? 'facter-ng' : 'facter'
```
Comment (ciprianbadescu) [inline lib/puppet.rb]: I don't know if it is cleaner/better/ruby-style, one alternative could be to have a facter-wrapper.rb with facter selection logic (currently `ENV['USE_FACTER_NG'] ? 'facter-ng' : 'facter'`) and require this `facter-wrapper` in puppet code

Comment (joshcooper) [reply lib/puppet.rb]: This is how we handled it previously when we switched from facter 2.x to cfacter: f41800ab4c0a928ff4f38ad5d1f6bd1e2b5619cd I think doing that is preferable as the implementation can be specified as a puppet setting, which is helpful when trying to run acceptance tests (since you only need to add the setting in a presuite step) and not every invocation of `puppet`

Comment (sebastian-miclea) [reply lib/puppet.rb]: @joshcooper I've updated the code today to use puppet settings, just trying to figure out why some specs fail.

File: lib/puppet/defaults.rb [hunk 4]
```diff
@@ -65,6 +65,25 @@ def self.default_vendormoduledir
 
   AS_DURATION = %q{This setting can be a time interval in seconds (30 or 30s), minutes (30m), hours (6h), days (2d), or years (5y).}
 
+  define_settings(:main,
+    :facterng => {
+        :default => false,
+        :type    => :boolean,
+        :desc    => 'Whether to enable a pre-Facter 4.0 release of Facter (distributed as
+          the "facter-ng" gem). This is not necessary if Facter 3.x or later is installed.
+          This setting is still experimental and has been only incuded on Windows builds',
```
Comment (joshcooper) [inline lib/puppet/defaults.rb]: nit typo `incuded`

File: man/man5/puppet.conf.5 [hunk 5]
```diff
@@ -344,6 +344,14 @@ The main Puppet configuration directory\. The default for this setting is calcul
...
 .IP "" 0
 .
+.SS "facterng"
+Whether to enable a pre-Facter 4.0 release of Facter in ruby\. Defaults to false\.
+.
+.IP "\(bu" 4
+\fIDefault\fR: false
+.
+.IP "" 0
+.
```
Comment (joshcooper) [inline man/man5/puppet.conf.5]: This file will be autogenerated on commit to puppet https://jenkins-master-prod-1.delivery.puppetlabs.net/view/puppet-agent/view/Acceptance%20Suites/view/master/view/Components/job/platform_puppet_vanagon-component-manpage-generation_master_agent-master/

**Comment by sebastian-miclea** [comment]
jenkins please test this on debian9-64a, debian10-64a, fedora29-64a, fedora30-64a, osx1014-64a, redhat5-64a, centos8-64a, redhat7-64a, redhat7-POWERa, sles12-64a, sles15-64a, solaris114-64a, ubuntu1604-64a, ubuntu1604-32a, , ubuntu1804-64a, windows10ent-64a, windows2012r2-64a, windows2016-64a, windows2019-64a
